### ISSUE TITLE
Loading the Inception_v4 model in IOS  and No OpKernel was registered to support Op error

### ISSUE THREAD

**Issue by vaklyuenkov** [tensorflow/models#790]
**Environment info**

iOS

**steps:**

1. I trained inception_v4 (slim) on a my dataset of 20 classes
2. Сonverted .ckpt into .pb by freeze_graph
3. Use .pb file, load into iOS and run

**Logs**

Could not create TensorFlow Graph: Invalid argument: No OpKernel was registered to support Op 'RandomUniform' with these attrs.  Registered devices: [CPU], Registered kernels: 
  <no registered kernels> 
 
  [[Node: InceptionV4/Logits/Dropout_1b/dropout/random_uniform/RandomUniform = RandomUniform[T=DT_INT32, dtype=DT_FLOAT, seed=0, seed2=0](InceptionV4/Logits/Dropout_1b/dropout/Shape)]] 
F /Users/pavelgerashchenko/Downloads/tensorflow-master-2/tensorflow/contrib/ios_examples/camera/CameraExampleViewController.mm:391] Couldn't load model: Invalid argument: No OpKernel was registered to support Op 'RandomUniform' with these attrs.  Registered devices: [CPU], Registered kernels: 
  <no registered kernels> 
 
  [[Node: InceptionV4/Logits/Dropout_1b/dropout/random_uniform/RandomUniform = RandomUniform[T=DT_INT32, dtype=DT_FLOAT, seed=0, seed2=0](InceptionV4/Logits/Dropout_1b/dropout/Shape)]]

Thank you.

**Comment by lciruelo** [comment]
Hi, how did you convert a .ckpt into .pb?

**Comment by vaklyuenkov** [comment]
there's the freeze_graph.py script that takes a graph definition and a set of checkpoints and freezes them together into a single file.

What this does is load the GraphDef, pull in the values for all the variables from the latest checkpoint file, and then replace each Variable op with a Const that has the numerical data for the weights stored in its attributes It then strips away all the extraneous nodes that aren't used for forward inference, and saves out the resulting GraphDef into an output file.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py

**Comment by lciruelo** [comment]
yeah!, I already knew that script, but the problem I have, It's that I dont know where I can find a graph definition

**Comment by vaklyuenkov** [comment]
Yeah) I rewrite a little this script for my goal. May be I miss something important # We retrieve our checkpoint fullpath
```
input_checkpoint = '..../v_43k.ckpt'

# We precise the file fullname of our freezed graph
output_graph = ".../frozen_model.pb"

# Before exporting our graph, we need to precise what is our output node
# This is how TF decides what part of the Graph he has to keep and what part it can dump
# NOTE: this variables is plural, because you can have multiple output nodes
output_node_names = "InceptionV4/Logits/Predictions"

# We clear the devices, to allow TensorFlow to control on the loading where it wants operations to be calculated
clear_devices = True

# We import the meta graph and retrive a Saver
saver = tf.train.import_meta_graph(input_checkpoint + '.meta', clear_devices=clear_devices)

# We retrieve the protobuf graph definition
graph = tf.get_default_graph()
input_graph_def = graph.as_graph_def()

# We start a session and restore the graph weights
with tf.Session() as sess:
        saver.restore(sess, input_checkpoint)

        # We use a built-in TF helper to export variables to constant
        output_graph_def = graph_util.convert_variables_to_constants(
            sess, # The session is used to retrieve the weights
            input_graph_def, # The graph_def is used to retrieve the nodes 
            output_node_names.split(",") # The output node names are used to select the usefull nodes
        )

        # Finally we serialize and dump the output graph to the filesystem
        with tf.gfile.GFile(output_graph, "wb") as f:
            f.write(output_graph_def.SerializeToString())
        print("%d ops in the final graph." % len(output_graph_def.node))

        tf.train.write_graph(output_graph_def, "..../", "frozen_model.pbtxt" , as_text=True)


```

**Comment by michaelisard** [comment]
Are you sure you are freezing the graph correctly? I am surprised that your inference graph contains a RandomUniform Op to do dropout: shouldn't dropout have been pruned away?

**Comment by lciruelo** [comment]
I dont know how to create the logits of inceptionV4 (output_node_names = "InceptionV4/Logits/Predictions"), any solution?

**Comment by jart** [comment]
This was most likely fixed in https://github.com/tensorflow/tensorflow/commit/35f8b1fe2c097ce97b5b465a67f312e92a62f162.

**Comment by cancan101** [comment]
I think this issues still exists. See: https://github.com/tensorflow/tensorflow/issues/6124 and https://github.com/tensorflow/tensorflow/issues/5867

**Comment by RootYuan** [comment]
"Сonverted .ckpt into .pb by freeze_graph"
@vaklyuenkov  , I'm sorry to bother you. when you did this , what "output_node_names" you used?
I don't know  what is the name of the output node. I'm using Inception_v4 too
